app/Http/Controllers/Auth/OAuth2Controller.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+namespace Pterodactyl\Http\Controllers\Auth;
+
+use Illuminate\Support\Arr;
+use Illuminate\Support\Str;
+use Illuminate\Auth\AuthManager;
+use Illuminate\Support\MessageBag;
+use Prologue\Alerts\AlertsMessageBag;
+use Illuminate\Contracts\Hashing\Hasher;
+use Laravel\Socialite\Facades\Socialite;
+use Pterodactyl\Http\Controllers\Controller;
+use Pterodactyl\Traits\Helpers\OAuth2Providers;
+use Pterodactyl\Services\Users\UserUpdateService;
+use Pterodactyl\Services\Users\UserCreationService;
+use Pterodactyl\Contracts\Repository\UserRepositoryInterface;
+
+class OAuth2Controller extends Controller
+{
+    use OAuth2Providers;
+
+    /**
+     * @var \Pterodactyl\Services\Users\UserCreationService
+     */
+    protected $creationService;
+
+    /**
+     * @var \Illuminate\Contracts\Hashing\Hasher
+     */
+    private $hasher;
+
+    /**
+     * @var \Illuminate\Auth\AuthManager
+     */
+    private $auth;
+
+    /**
+     * @var \Pterodactyl\Contracts\Repository\UserRepositoryInterface
+     */
+    protected $repository;
+
+    /**
+     * @var \Pterodactyl\Services\Users\UserUpdateService
+     */
+    protected $updateService;
+
+    /**
+     * @var \Prologue\Alerts\AlertsMessageBag
+     */
+    protected $alert;
+
+    /**
+     * OAuth2Controller constructor.
+     *
+     * @param \Pterodactyl\Services\Users\UserCreationService $creationService
+     * @param \Illuminate\Contracts\Hashing\Hasher $hasher
+     * @param \Illuminate\Auth\AuthManager $auth
+     * @param \Pterodactyl\Contracts\Repository\UserRepositoryInterface $repository
+     * @param \Pterodactyl\Services\Users\UserUpdateService $updateService
+     * @param AlertsMessageBag $alert
+     */
+    public function __construct(UserCreationService $creationService, Hasher $hasher, AuthManager $auth, UserRepositoryInterface $repository, UserUpdateService $updateService, AlertsMessageBag $alert)
+    {
+        $this->creationService = $creationService;
+        $this->hasher = $hasher;
+        $this->auth = $auth;
+        $this->repository = $repository;
+        $this->updateService = $updateService;
+        $this->alert = $alert;
+    }
+
+    public function login()
+    {
+        // If this feature is not enabled or the user is already signed in and isn't linking a new account redirect or abort
+        if ($this->auth->guard()->check() and ! session()->has('link_oauth2_driver')) {
```

Comment:
Please try to avoid using these global functions where possible. I know `view` and `redirect` get used a bunch, but you don't need `session()` or `abort()`. You already have the request object available here, and you can throw `NotFoundHttpException`s to cause 404s. 👍